Applications/Spire/Source/Ui/DecimalInputWidget.cpp

```diff
@@ -0,0 +1,269 @@
+#include "Spire/Ui/DecimalInputWidget.hpp"
+#include <QFocusEvent>
+#include <QKeyEvent>
+#include <QLineEdit>
+#include <QMouseEvent>
+#include "Spire/Spire/Dimensions.hpp"
+
+using namespace boost::signals2;
+using namespace Spire;
+
+namespace {
+  const auto DEFAULT_DECIMAL_PLACES = 6;
+  const auto SHIFT_STEPS = 10;
+  const auto DOWN_ARROW_ICON = ":/Icons/arrow-down.svg";
+  const auto DOWN_ARROW_HOVER_ICON = ":/Icons/arrow-down-hover.svg";
+  const auto DOWN_ARROW_DISABLED_ICON = ":/Icons/arrow-down-disabled.svg";
+  const auto UP_ARROW_ICON = ":/Icons/arrow-up.svg";
+  const auto UP_ARROW_HOVER_ICON = ":/Icons/arrow-up-hover.svg";
+  const auto UP_ARROW_DISABLED_ICON = ":/Icons/arrow-up-disabled.svg";
+
+  auto ARROW_WIDTH() {
+    static auto width = scale_width(8);
+    return width;
+  }
+
+  auto BUTTON_PADDING() {
+    static auto padding = scale_width(15);
+    return padding;
+  }
+}
+
+DecimalInputWidget::DecimalInputWidget(double value, QWidget* parent)
+    : QDoubleSpinBox(parent),
+      m_last_valid_value(value),
+      m_last_cursor_pos(0),
+      m_has_first_click(false) {
+  setValue(m_last_valid_value);
+  setDecimals(DEFAULT_DECIMAL_PLACES);
+  set_stylesheet(false, false);
+  setContextMenuPolicy(Qt::NoContextMenu);
+  connect(this, QOverload<double>::of(&QDoubleSpinBox::valueChanged), this,
+    &DecimalInputWidget::on_value_changed);
+  connect(lineEdit(), &QLineEdit::textEdited, this,
+    &DecimalInputWidget::on_text_edited);
+  lineEdit()->installEventFilter(this);
+}
+
+QString DecimalInputWidget::textFromValue(double value) const {
```

Comment:
This method does not use `this` and can be static.

Reply:
The textFromValue() method is overriding a base class method. I did change it to be protected to match the base class' declaration.